hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/impl/BlockManagerImpl.java

```diff
@@ -343,7 +343,7 @@ public List<BlockData> listBlock(Container container, long startLocalID, int
                 .getSequentialRangeKVs(startKey, count,
                     cData.containerPrefix(), cData.getUnprefixedKeyFilter());
         for (Table.KeyValue<String, BlockData> entry : range) {
-          result.add(entry.getValue());
+          result.add(db.getStore().getBlockByID(null, entry.getKey()));
```

Comment:
If incremental chunk list feature is enabled, the last chunk of a block may exist in the lastChunkInfoTable. So getBlockByID() is used here to get the complete block data.

There could be opportunities to optimize later.

Reply:
Can we put the merge of last chunk into another function? For example:
```
  @Override
  public BlockData getBlockByID(BlockID blockID,
      String blockKey) throws IOException {
    // check block data table
    BlockData blockData = getBlockDataTable().get(blockKey);
    return getCompleteBlockData(blockData, blockID, blockKey);
  }

  @Override
  public BlockData getCompleteBlockData(
      BlockData blockData, BlockID blockID, String blockKey) throws IOException {
    BlockData lastChunk = null;
    if (blockData == null || isPartialChunkList(blockData)) {
      // check last chunk table
      lastChunk = getLastChunkInfoTable().get(blockKey);
    }

    if (blockData == null || blockData.getChunks().isEmpty()) {
      if (lastChunk == null) {
        throw new StorageContainerException(
            NO_SUCH_BLOCK_ERR_MSG + " BlockID : " + blockID, NO_SUCH_BLOCK);
      } else {
        if (LOG.isDebugEnabled()) {
          LOG.debug("blockData=(null), lastChunk={}", lastChunk.getChunks());
        }
        return lastChunk;
      }
    } else {
      if (lastChunk != null) {
        reconcilePartialChunks(lastChunk, blockData);
      } else {
        if (LOG.isDebugEnabled()) {
          LOG.debug("blockData={}, lastChunk=(null)", blockData.getChunks());
        }
      }
    }

    return blockData;
  }
```
Then, we can use it here:
```
for (Table.KeyValue<String, BlockData> entry : range) {
          result.add(db.getStore().getCompleteBlockData(entry.getValue(), null, entry.getKey()));
        }
        return result;
```

I think it is a simple change to avoid getting blockData from table twice.

Reply:
Thanks! That makes sense to me.